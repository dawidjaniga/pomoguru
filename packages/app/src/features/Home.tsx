import React, { useEffect } from 'react'

import Button from 'core/components/Button'
import { Link } from 'react-router-dom'
import WindowContent from 'core/components/WindowContent'
import WindowTitle from 'core/components/WindowTitle'
import format from 'date-fns/format'
import slack from 'api/slack'
import styled from 'styled-components'
import { useMediator } from 'core/mediator'
import { useTimer } from 'core/timer'

const CloseLink = styled.a`
  color: var(--text-color);
`

const Buttons = styled.div`
  display: flex;
  justify-content: center;
  width: 46%;
`

const Timer = styled.div`
  color: var(--text-color);
  font-size: 58px;
  font-weight: 600;
  line-height: 79px;
  text-align: center;
`

export default function Home () {
  const { onInit, onStartClick, onPauseClick, onStopClick } = useMediator()
  const [{ timeLeft, isActive }] = useTimer()

  useEffect(() => {
    onInit()
  }, [])

  return (
    <>
      <WindowTitle>
        <div>
          PomoGuru <img src='icon-small.png' alt='PomoGuru icon' />
        </div>
        <CloseLink>x</CloseLink>
      </WindowTitle>
      <WindowContent>
        {slack.isInstalled() ? (
          <>
            <Timer data-testid='timer'>
              {format(new Date(timeLeft * 1000), 'mm:ss')}
            </Timer>
            <Buttons>
              {isActive ? (
                <>
                  <Button onClick={onPauseClick} data-testid='pause'>
                    Pause
                  </Button>
                  <Button onClick={onStopClick} data-testid='stop'>
                    Stop
                  </Button>
                </>
              ) : (
                <Button onClick={onStartClick} data-testid='start'>
                  Start
                </Button>
              )}
            </Buttons>
          </>
        ) : (
          <Link to='/slack-installation' data-testid='connect-to-slack'>
            Connect to Slack
          </Link>
        )}
      </WindowContent>
    </>
  )
}
