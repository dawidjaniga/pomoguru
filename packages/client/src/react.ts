import { useCaseProvider, UseCaseNames, model } from './index'
import { useCallback, useEffect, useState } from 'react'
import { Publisher } from './objects/publisher'
import { UseCaseInput, UseCaseOutput } from './core/useCasesMap'

export function useUser () {
  const [user, setUser] = useState(model.get('user'))

  useEffect(() => {
    model.subscribe('user:changed', (value: unknown) => {
      setUser(value)
    })
  }, [])

  return user
}

export function useNotificationsAllowed () {
  const [notificationsAllowed, setNotificationsAllowed] = useState(
    model.get('notificationsAllowed')
  )

  useEffect(() => {
    model.subscribe('notificationsAllowed:changed', (value: boolean) => {
      setNotificationsAllowed(value)
    })
  }, [])

  return notificationsAllowed
}

export function useCase<Name extends UseCaseNames> (
  useCase: Name,
  options?: UseCaseInput<Name>
  // @Improvement: Allow useCase to pass options with correct type
) {
  const concreteUseCase = useCaseProvider.get(useCase)
  const [loaded, setIsLoaded] = useState<boolean>(false)
  const [error, setError] = useState<Error>()
  const [data, setData] = useState<UseCaseOutput<Name>>()
  // @TODO: #improvement
  // Return correct Output type
  // const [data, setData] = useState<
  //   Awaited<ReturnType<typeof concreteUseCase['execute']>>
  // >()

  const getData = useCallback(async function () {
    /*
    @TODO: #improvement
    Allow to pass options to UseCase. Altering behavior while executing or by setters?
    */

    const data = await concreteUseCase.execute()

    // @ts-ignore
    setData(data)
    setIsLoaded(true)
    // try {
    // } catch (e) {
    //   throw e
    //   if (e instanceof Error) {
    //     setError(e)
    //     setIsLoaded(true)
    //   }
    // }
  }, [])

  useEffect(() => {
    getData()

    if (concreteUseCase instanceof Publisher) {
      concreteUseCase.subscribe('updated', () => {
        getData()
      })
    }
  }, [])
  // }, [JSON.stringify(options)])

  return {
    loaded,
    error,
    data
  }
}
