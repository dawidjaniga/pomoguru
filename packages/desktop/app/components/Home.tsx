import React from 'react';
import SlackService from './../services/slack'
import LocalStoreService from './../services/localStore'

export default function Home() {
  return (
    <div data-tid="container">
      <h2>Home</h2>
      <a onClick={SlackService.goIntoFocus}>Go to focus</a>
      <a onClick={SlackService.endFocus}>End focus</a>
      <a onClick={() => LocalStoreService.set('slack-token', '')}>Reset token</a>

    </div>
  );
}
