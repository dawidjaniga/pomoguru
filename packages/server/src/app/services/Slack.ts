import getUnixTime from 'date-fns/getUnixTime'
import { ApplicationError } from '@server/types/errors'
import axios from 'axios'
import qs from 'qs'

type AuthorizeResponse = {
  ok: boolean
  error: string
  authed_user: {
    access_token: string
  }
}

type ProfileResponse = {
  ok: boolean
  error: string
}
interface Status {
  text: string
  emoji: string
  expireTime: Date | number
}

export class Slack {
  static async authorize (code: string): Promise<string> {
    const { data } = await axios.post<AuthorizeResponse>(
      'https://slack.com/api/oauth.v2.access',
      qs.stringify({
        client_id: process.env.SLACK_CLIENT_ID,
        client_secret: process.env.SLACK_CLIENT_SECRET,
        code,
        redirect_uri: 'https://localhost:4200/slack'
      }),
      {
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded'
        }
      }
    )

    if (data.ok && data.authed_user.access_token) {
      return data.authed_user.access_token
    } else {
      throw new ApplicationError(`Slack oauth error: ${data.error}`)
    }
  }

  static async changeStatus (
    { text, emoji, expireTime }: Status,
    userToken: string
  ) {
    const payload = qs.stringify({
      token: userToken,
      profile: JSON.stringify({
        status_text: text,
        status_emoji: emoji,
        status_expiration: getUnixTime(expireTime)
      })
    })
    const { data } = await axios.post<ProfileResponse>(
      'https://slack.com/api/users.profile.set',
      payload,
      {
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded'
        }
      }
    )

    console.log('payload', payload, data)

    if (!data.ok) {
      throw new ApplicationError(`Slack Profile Set error: ${data.error}`)
    }
  }
}
