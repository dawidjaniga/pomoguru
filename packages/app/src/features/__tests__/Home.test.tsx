import React from 'react'
import { Route, BrowserRouter as Router } from 'react-router-dom'
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import Home from './../Home'
import Slack from 'api/slack'

jest.mock('electron', () => ({
  ipcRenderer: {
    send: jest.fn()
  }
}))

const onInitMock = jest.fn()
const onStartClickMock = jest.fn()
const onPauseClickMock = jest.fn()
const onStopClickMock = jest.fn()

jest.mock('core/mediator', () => ({
  useMediator: () => ({
    onInit: onInitMock,
    onStartClick: onStartClickMock,
    onPauseClick: onPauseClickMock,
    onStopClick: onStopClickMock
  })
}))

const timerMinutes = 10
const secondsInMinutes = 60
let timeLeft = secondsInMinutes * timerMinutes
let isActive: boolean

jest.mock('core/timer', () => ({
  useTimer: () => [
    {
      isActive,
      timeLeft
    }
  ]
}))

describe('Home View', () => {
  it('should run Mediator.onInit when loaded', () => {
    render(
      <Router>
        <Route path='/'>
          <Home />
        </Route>
      </Router>
    )

    expect(onInitMock).toBeCalled()
  })

  describe('when Slack is not initialized', () => {
    beforeEach(() => {
      Slack.isInstalled = () => false
    })

    it('should show "Connect to Slack" button', () => {
      const { container } = render(
        <Router>
          <Route path='/'>
            <Home />
          </Route>
        </Router>
      )

      expect(screen.getByTestId('connect-to-slack')).toBeInTheDocument()
      expect(container).toMatchSnapshot()
    })
  })

  describe('when Slack is initialzied', () => {
    beforeEach(() => {
      Slack.isInstalled = () => true
    })

    describe('when Timer is not active', () => {
      beforeEach(() => {
        isActive = false
      })

      it('should display: Timer in minutes, Start button and hide "Connect to Slack"', () => {
        render(
          <Router>
            <Route path='/'>
              <Home />
            </Route>
          </Router>
        )

        expect(screen.getByTestId('timer')).toHaveTextContent(
          timerMinutes + ':00'
        )
        expect(screen.getByTestId('start')).toBeInTheDocument()
        expect(screen.queryByTestId('connect-to-slack')).toBeNull()
      })

      it('should have onStartClick attached', () => {
        render(
          <Router>
            <Route path='/'>
              <Home />
            </Route>
          </Router>
        )

        userEvent.click(screen.getByTestId('start'))
        expect(onStartClickMock).toBeCalled()
      })
    })

    describe('when Timer is active', () => {
      beforeEach(() => {
        isActive = true
      })

      it('should have onPauseClick and onStopClick attached  ', () => {
        render(
          <Router>
            <Route path='/'>
              <Home />
            </Route>
          </Router>
        )

        userEvent.click(screen.getByTestId('pause'))
        userEvent.click(screen.getByTestId('stop'))
        expect(onPauseClickMock).toBeCalled()
        expect(onStopClickMock).toBeCalled()
      })
    })
  })
})
