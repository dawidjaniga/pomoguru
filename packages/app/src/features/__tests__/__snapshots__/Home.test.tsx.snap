// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home View when Slack is not initialized should show "Connect to Slack" button 1`] = `
<div>
  <div
    class="sc-dlfnbm lnEtAU"
  >
    <div>
      PomoGuru 
      <img
        alt="PomoGuru icon"
        src="icon-small.png"
      />
    </div>
    <a
      class="sc-hKgILt dXmNWo"
    >
      x
    </a>
  </div>
  <div
    class="sc-gsTCUz imZhzo"
  >
    <a
      data-testid="connect-to-slack"
      href="/slack-installation"
    >
      Connect to Slack
    </a>
  </div>
</div>
`;
